### PR TITLE
Merge paramterized u1, u2 and p gate in Optimize1qGates

### PR DESCRIPTION
**Pull request by rafal-pracht**
### Summary

Before the fix, the 'Optimize1qGates' transpiler optimization join together the gate only when the parameter was bounded (or when there was no parameter at all). After the fix, the u1, u2, and p gates will be combined even when the parameter is unbounded.

from qiskit import QuantumCircuit
from qiskit.circuit import Parameter
from qiskit.transpiler import PassManager
from qiskit.transpiler.passes import Optimize1qGates, Unroller

phi = Parameter('φ')
alpha = Parameter('α')

qc = QuantumCircuit(1)
qc.u1(2*phi, 0)
qc.u1(alpha, 0)
qc.u1(0.1, 0)
qc.u1(0.2, 0)
qc.draw(output='mpl')

image

pm = PassManager([Unroller(['u1', 'cx']), Optimize1qGates()])
nqc = pm.run(qc)
nqc.draw(output='mpl')

This change is a part of the advocate project done under @nbronn.

image

This is continuation of the issue https://github.com/Qiskit/qiskit-terra/pull/7309

### CODE REVIEW

**Review by mtreinish — COMMENTED** [review]
I'm a bit confused by this PR, it looks identical to #7309 except you've added the release note I asked for in my review yesterday. To add the release note there is no need to open a new pull request you can just modify the existing branch for #7309. It's generally better to use the existing PR so the entire review history is present and the merged commit refers to the single place where all the review happened.

**Comment by rafal-pracht** [comment]
@mtreinish you are absolutely right. I'm unable to add release notes to PR because I'm also working on the @nbronn  fork of terra and GitHub lost the connection between my code and the last PR. Therefore I've created a new one with exactly the same code with the release notes added.

Sorry for confusion.

File: releasenotes/notes/optimization-u2-gates-with-parameters-322b6c523251108c.yaml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+      qc.u1(alpha, 0)
+      qc.u1(0.1, 0)
+      qc.u1(0.2, 0)
+
+      pm = PassManager([Unroller(['u1', 'cx']), Optimize1qGates()])
+      nqc = pm.run(qc)
+
+    will be combined to the circuit with only one single-qubit gate
+      qc = QuantumCircuit(1)
+      qc.u1(2*phi + alpha + 0.3, 0)
```
Comment (mtreinish) [inline releasenotes/notes/optimization-u2-gates-with-parameters-322b6c523251108c.yaml:26]:
```suggestion
    The :class:`.Optimize1qGates` transpiler pass now has support for optimizing :class:`.U1Gate`,
    :class:`.U2Gate`, and :class:`.PhaseGate` gates with unbound parameters in a circuit.
    Previously, if these gates had unbound parameters the pass would not use them. For example::

        from qiskit import QuantumCircuit
        from qiskit.circuit import Parameter
        from qiskit.transpiler import PassManager
        from qiskit.transpiler.passes import Optimize1qGates, Unroller

        phi = Parameter('φ')
        alpha = Parameter('α')

        qc = QuantumCircuit(1)
        qc.u1(2*phi, 0)
        qc.u1(alpha, 0)
        qc.u1(0.1, 0)
        qc.u1(0.2, 0)

        pm = PassManager([Unroller(['u1', 'cx']), Optimize1qGates()])
        nqc = pm.run(qc)

    will be combined to the circuit with only one single-qubit gate::
    
        qc = QuantumCircuit(1)
        qc.u1(2*phi + alpha + 0.3, 0)
```

File: qiskit/transpiler/passes/optimization/optimize_1q_gates.py [hunk 2]
```diff
@@ -335,7 +337,7 @@ def _split_runs_on_parameters(runs):
 
     out = []
     for run in runs:
-        groups = groupby(run, lambda x: x.op.is_parameterized())
+        groups = groupby(run, lambda x: x.op.is_parameterized() and x.op.name == "u3")
```
Comment (mtreinish) [inline qiskit/transpiler/passes/optimization/optimize_1q_gates.py:340]: Don't we need to handle `u` here too, not just `u3`?

File: qiskit/transpiler/passes/optimization/optimize_1q_gates.py [hunk 3]
```diff
@@ -97,9 +98,6 @@ def run(self, dag):
                     and current_node.op.definition.global_phase
                 ):
                     right_global_phase += current_node.op.definition.global_phase
-                # If there are any sympy objects coming from the gate convert
-                # to numpy.
-                left_parameters = tuple(float(x) for x in left_parameters)
```
Comment (mtreinish) [inline qiskit/transpiler/passes/optimization/optimize_1q_gates.py:102]: Per the other PR, shouldn't this be:

```python
                try:
                    left_parameters = tuple(float(x) for x in left_parameters)
                except TypeError:
                    # If left_parameters contained any unbound Parameters
                    pass
```

**Review by mtreinish — COMMENTED** [review]
It looks like some of the revisions made on #7309 didn't make it through the transition to this new PR. I also left some inline improvement suggestions on the release note. Mostly around som sphinx rst syntax fixes and some mild rewording.

**Comment by rafal-pracht** [comment]
@mtreinish I've added those changes.

**Review by mtreinish — APPROVED** [review]
LGTM, thanks for keeping with this and making the updates

**Pull request merged**